Bylaws Overhaul

TinoDidriksen opened the pull request:
Easier to track changes with a PR.

Goals of the changes:
- Prepare for creating a legal entity for Apertium. But that will come later; this overhaul is needed regardless.
- Rename Developer to Contributor.
- Rename Committer to Voting Member. Not merely Member, as that would be confused for member of the Github organisation. Anyone can be a Github member, but it has to be a conscious choice to want to be a voting member.
- Separate the definition of Contributor with commit access and Voting Member. Being a voter should not be tied to having commit rights. This tie made sense with a monorepo, but it really doesn't make sense with split decentralized repos.
- Empower Contributors to do what they're effectively doing anyway (make releases, grant access), while still keeping PMC oversight for big changes.
- Recommend a Contributor License Agreement, but at least require that licenses must be clear. I'm tired of resolving licensing issues.
- Raise inactivity limit to 1 year.
- Formally ensure we reduce the bus factor and security risks.
- Change PMC voting system to STV.
- Formalize tie-breaker mechanics, since it just became relevant.

@mlforcada @ftyers @xavivars @jonorthwash @sushain97

jonorthwash commented:
> PMC votes have traditionally been very fast (four email votes and done).

Is PMC voting as simple majority specified anywhere?

TinoDidriksen commented:
> Is PMC voting as simple majority specified anywhere?

Not explicitly, but "The Project Management Committee votes on all of its decisions. The President has the casting vote." implies it.

But not sure I'd want it explicitly stated, because that would slow down work. We currently operate a lot on lack of dissent. Majority consenting votes for everything would be nice, but unrealistic.

As long as decisions are actually filed somewhere so that they can be reviewed, lack of dissent ought to do.

TinoDidriksen commented:
Pushed new commit, which further clean up wording. I think this resolves all current comments, minus those below.

----

> I'd change twelve to twenty-four, as someone mentioned in an email thread. That would make it "a whole eleccion cycle" (if we keep the schedules for those).

A reason against this is that other votes may come up between elections, and we should want only active members to vote on these. I think a year is a reasonable balance between administrative overhead and engagement.

----

@khannatanmai brought up a point about majority voting being potentially bad for 3-4 splits, which has some merit. If that many dissent, maybe we should pause a decision. We could add a clause saying that 3 PMC members can require asking the whole Assembly.

jonorthwash commented:
> @khannatanmai brought up a point about majority voting being potentially bad for 3-4 splits, which has some merit. If that many dissent, maybe we should pause a decision. We could add a clause saying that 3 PMC members can require asking the whole Assembly.

It would be nice to have a model of decisions being made only when there's "consensus", whatever we interpret that to mean.  But I think that would slow things down.  Perhaps the reuslts of a majority Assembly vote can be used as a single PMC vote sometimes?

Worst case scenario, you have a PMC and Assembly that are both entirely split, and if the majority that "wins" a vote is only a very small majority, that could result in a split into two projects.  The question of who the accounts follow and the like would probably be the ones who don't "resign" or leave, so probably the majority ...

Anyway, hopefully it never comes to that.

xavivars commented:
> A reason against this is that other votes may come up between elections, and we should want only active members to vote on these. I think a year is a reasonable balance between administrative overhead and engagement

I still think there's some value in the the 24 months: in an election, anyone who has been an active contributor since last election will have voting rights. But don't feel strongly about it.

khannatanmai commented:
> It would be nice to have a model of decisions being made only when there's "consensus", whatever we interpret that to mean. But I think that would slow things down. Perhaps the reuslts of a majority Assembly vote can be used as a single PMC vote sometimes?

A consensus amongst seven people from the same organisation doesn't seem that unreasonable to me. However, dissent isn't uncommon, which is why we could even aim for a system where 1 person dissenting is acceptable, but 2 people dissenting is reasonable doubt, and causes the motion to fall.

While most decisions won't reach such a case since a vote would be done after thorough discussion, in the worst case scenario, such a method would help.

> Worst case scenario, you have a PMC and Assembly that are both entirely split, and if the majority that "wins" a vote is only a very small majority, that could result in a split into two projects. The question of who the accounts follow and the like would probably be the ones who don't "resign" or leave, so probably the majority ...

A decision that is supported by only a very small majority after thorough discussion is bad news. These 7 people that have been elected and are the project management committee, each capable in their own right, and if a situation should arrive where there's a significant split in the PMC and the Assembly, the best course of action is to not go ahead with that decision, since there's reasonable doubt (which still needs to be resolved). We can think of it as a big minority veto if needed.

khannatanmai commented:
Also unrelated: STV is a great method of voting and very rarely has ties as if there's a tie when counting the voters' first rank votes, we move to the second rank, then third, and so on. There is still an almost negligible possibility of a tie for which we might wanna have a structure in place but overall STV will bring tons of benefits.

TinoDidriksen commented:
> I still think there's some value in the the 24 months: in an election, anyone who has been an active contributor since last election will have voting rights. But don't feel strongly about it.

Eh, I don't mind either, really. As long as there is an expiry mechanism. It's not as if we've ever enforced the current 6 month limit, so this would be the first time it gets enforced.

Which reminds me, there was also discussion about a forced removal mechanism. We currently have no formal way to rescind voting rights in case of abuse/misbehavior. Should add a clause that lets the Assembly vote to remove a member, if prompted to by the PMC.

jonorthwash commented:
> Which reminds me, there was also discussion about a forced removal mechanism. We currently have no formal way to rescind voting rights in case of abuse/misbehavior. Should add a clause that lets the Assembly vote to remove a member, if prompted to by the PMC.

I think this is a good idea.

TinoDidriksen commented:
Pushed another commit. Please stop commenting on the original commit - it is now rather out of date.

Added minority consensus escape hatch. Added voting member removal mechanism.

----

> My inclination would be for Apertium to encourage repos to be community developed under the org account as easily as possible. Having a repo on GitHub is more about development of a product (a work in progress) than about distributing a product ("finished" work).

Fixed? I've split up what requires majority vs. what requires single person approval. This probably needs tuning, or more subclauses added.

There needs to be a balance between protecting the Apertium brand and smooth development. With this split, a majority consensus is needed to ship something as an official Apertium product, but development can start with just one member signing off on it.

----

> This is why I liked "committers". How about "Apertium developers"? That's what they are!

No! A major part of this is that contributors are not necessarily developers. Even if you don't commit a single line of code, you can still be an active contributor. The original bylaws spelled this out clearly enough, but used the word developer. If someone is active on the mailing list, helping end-users, or writing papers, that should naturally count.

I really want that split cleanly. Status should not be tied to code contributions, but to contributions in any form.

----

> Seniority would have to be unambiguously defined, and any way I can think of is quite hard to determine...

Really? In Danish law this is quite clearly defined as the person who's been attached to the company the longest. Which given the breakdown, it would be nigh impossible to wind up in a situation where the tie is between two with the same seniority.

jonorthwash commented:
> PMC votes have usually taken around 1 day.

It's been rare for the entire PMC to respond to a vote.  Maybe that'll be different with the new PMC.

> Commit permissions should be granted by the person in charge of the corresponding repository, then reported to the PMC for ratification. The PMC would just ratify.
> Xavi is right: for more substantial repository changes like the ones he describes, the PMC should be involved.

How should these be coded up?

jonorthwash commented:
> There needs to be a balance between protecting the Apertium brand and smooth development. With this split, a majority consensus is needed to ship something as an official Apertium product, but development can start with just one member signing off on it.

I'm happy with this.

khannatanmai commented:
> I am not convinced of the need for a STV system. What makes it so substantially better than the one we have now?

@mlforcada Apart from a tie breaker, the USP of an STV system is that unlike the current system, which is in effect - I want these people to be a part of the PMC and the others I don't care about, it gives the voter an ability to rank people, so that their vote isn't just about who should be a part of the PMC but who according to them deserves it more. This helps not only in tie breaking, but in the event of a large number of candidates it provides a much more fair system of distributing votes.

TinoDidriksen commented:
> > Commit permissions should be granted by the person in charge of the corresponding repository, then reported to the PMC for ratification. The PMC would just ratify.
> > Xavi is right: for more substantial repository changes like the ones he describes, the PMC should be involved.
> 
> How should these be coded up?

Current wording provides that, I would say.

But this is where interpretation of the bylaws comes in. My interpretation is that substantial changes requires a PMC member to sign off on it, which constitutes a PMC decision. And PMC decisions must be filed with the rest of the PMC and eventually the Assembly.

I've been thinking we should have an accompanying "The Spirit of" document to write the long form interpretation and rationale of each clause.

scoopgracie commented:
1. I oppose allowing the Secretary and Treasurer to be the same person.

2. Using additional criteria for tiebreaking could easily turn into a rigged election.

3. I believe there should be a court of some sort for handling violations. See the Bylaw Violation Court in http://wiki.apertium.org/wiki/User:ScoopGracie/PMC/Proposed_bylaws for an example

scoopgracie commented:
I also think there should be a way for the PMC proposals to organized in a form of legal code.

scoopgracie commented:
I strongly support the STV system.

TinoDidriksen commented:
>     1. I oppose allowing the Secretary and Treasurer to be the same person.

On what grounds? There is no formal non-profit organisation requirements that we even need a separate treasurer or secretary, and historically we haven't practically had either.

>     2. Using additional criteria for tiebreaking could easily turn into a rigged election.

How? The tie-breaker doesn't even become relevant unless there is a tie, which with STV is very unlikely.

>     3. I believe there should be a court of some sort for handling violations. See the Bylaw Violation Court in [http://wiki.apertium.org/wiki/User:ScoopGracie/PMC/Proposed_bylaws for](http://wiki.apertium.org/wiki/User:ScoopGracie/PMC/Proposed_bylaws%C2%A0for) an example

That is effectively what I put in there. The PMC can quickly suspend where needed, and then must refer the matter to the Assembly for a final decision. A more specific mechanism is not needed.

scoopgracie commented:
Okay. #s 1 and 2 make sense. On # 3, I think a non PMC court should decide
removals for separation of powers.



On Mon, Apr 20, 2020, 22:14 Tino Didriksen <notifications@github.com> wrote:

> 1. I oppose allowing the Secretary and Treasurer to be the same person.
>
> On what grounds? There is no formal non-profit organisation requirements
> that we even need a separate treasurer or secretary, and historically we
> haven't practically had either.
>
> 2. Using additional criteria for tiebreaking could easily turn into a rigged election.
>
> How? The tie-breaker doesn't even become relevant unless there is a tie,
> which with STV is very unlikely.
>
> 3. I believe there should be a court of some sort for handling violations. See the Bylaw Violation Court in [http://wiki.apertium.org/wiki/User:ScoopGracie/PMC/Proposed_bylaws for](http://wiki.apertium.org/wiki/User:ScoopGracie/PMC/Proposed_bylaws%C2%A0for) an example
>
> That is effectively what I put in there. The PMC can quickly suspend where
> needed, and then must refer the matter to the Assembly for a final
> decision. A more specific mechanism is not needed.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/apertium/organisation/pull/13#issuecomment-616956548>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AN54XPZONJHWVI3WEONJV5TRNUTT3ANCNFSM4MH2I5SQ>
> .
>

TinoDidriksen commented:
> On # 3, I think a non PMC court should decide removals for separation of powers.

Yes, that is the way it is written. The PMC can temporarily remove access, but it is up to the Assembly to actually strip voting rights permanently. But the PMC must be able to temporarily revoke access immediately, to prevent abuse.

scoopgracie commented:
I understand, but I think the removal power (as well as power to interpret
the bylaws) should be held by some elected or appointed group other than
the PMC.

On Mon, Apr 20, 2020, 22:23 Tino Didriksen <notifications@github.com> wrote:

> On # 3, I think a non PMC court should decide removals for separation of
> powers.
>
> Yes, that is the way it is written. The PMC can temporarily remove access,
> but it is up to the Assembly to actually strip voting rights permanently.
> But the PMC must be able to temporarily revoke access immediately, to
> prevent abuse.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/apertium/organisation/pull/13#issuecomment-616959248>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AN54XP52XXBFG6OPRUJ7XMLRNUUWNANCNFSM4MH2I5SQ>
> .
>

TinoDidriksen commented:
> I understand, but I think the removal power (as well as power to interpret the bylaws) should be held by some elected or appointed group other than the PMC.

It already is. The Assembly holds all the power, for everything. The PMC is just executive. This has always been the case.

And the proposed bylaws protects that even more, by requiring the whole PMC to have all the keys, not just single members.

ftyers commented on Bylaws.md:
```diff
@@ -3,108 +3,132 @@
 </ol>
 
-## Roles and responsibilities
+## Roles and Responsibilities
```
Why change to title case? I prefer sentence case for headings.

ftyers commented on Bylaws.md:
```diff
@@ -3,108 +3,132 @@
 
 <ol start="4">
-	<li>Developers are  volunteers who are contributing time, code, documentation, or resources to Apertium. A developer that makes sustained, welcomed contributions to the project may be invited to become a committer, though the exact timing of such invitations depends on many factors</li>
+	<li>Contributors are volunteers who contribute time, code, documentation, or resources to Apertium. A contributor that makes sustained, welcomed contributions to Apertium may be invited to become a Voting Member, though the exact timing of such invitations depends on many factors.</li>
```
"many factors" is vague. maybe say "the exact timing is not specified"

ftyers commented on Bylaws.md:
```diff
@@ -3,108 +3,132 @@
-	<li>Developers are  volunteers who are contributing time, code, documentation, or resources to Apertium. A developer that makes sustained, welcomed contributions to the project may be invited to become a committer, though the exact timing of such invitations depends on many factors</li>
+	<li>Contributors are volunteers who contribute time, code, documentation, or resources to Apertium. A contributor that makes sustained, welcomed contributions to Apertium may be invited to become a Voting Member, though the exact timing of such invitations depends on many factors.</li>
+	<li>Contributors who have administrative rights to a repository may exercise those rights almost fully, including granting rights to new contributors, making releases, or setting up automation. Potentially destructive or distruptive actions, such as deleting or renaming the repository or master branch, or breaking backwards compatibility, must be approved by a PMC member, who may decide the acti ...
+	<li>Apertium recommends, but does not require, that contributors sign the Apertium Open Source Contributor License Agreement. However, contributions that are not licensed to Apertium in this way must be clearly marked with a GPLv2+ compatible license, plus the author's name and email or other identifying information sufficient for a lawyer to reach the contributor.</li>
```
Free/Open-Source ? How about other contributions under CC-BY(-SA)?

TinoDidriksen replied:
CC-BY is directly GPLv2+ compatible. CC-BY-SA 4.0 is GPLv3+ compatible, so that falls under the +. Too much of our other stuff is GPLv2+ and almost certainly will not be relicensed under anything else, so I see no other solution than to make sure anything new is up front compatible with GPLv2+. But we only need compatible - and many licenses are compatible. It is a very minor restriction.
What I would really want is copyright assignment from everyone, but I think that will be impossible to push through.

ftyers replied:
Works for me, how about calling it Free/Open-Source though? @mlforcada what do you think?

TinoDidriksen replied:
Oh, you want to change the name of the CLA? Sure. Doesn't matter what it is called - it's the clauses in it that matter.

ftyers commented on Bylaws.md:
```diff
@@ -3,108 +3,132 @@
+## Transitional provisions
+
+<ol start="31">
+	<li>The initial Assembly of Voting Members consists of everyone who were eligible to vote in the March 2020 PMC election, minus those who have been inactive for over two years. Voting in the election shall count as being active for this purpose.</li>
```
who was

ftyers commented on CLA-optional.md:
```diff
@@ -0,0 +1,20 @@
+- **Project** - is an umbrella term that refers to any and all Apertium open source projects.
+- **Contribution** - any type of work that is submitted to a Project, including any modifications or additions to existing work.
+- **Submitted** - conveyed to a Project via a pull request, commit, issue, or any form of electronic, written, or verbal communication with Apertium, contributors or maintainers.
+- **Open Source License** - any license that matches the Open Source Initiative's The Open Source Definition and/or Free Software Foundation's The Free Software Definition. See https://opensource.org/docs/osd and https://gnu.org/philosophy/free-sw.html
```
Consistent spelling "licence"

mr-martian commented:
During the most recent election, the election board agreed that since we could see all the ballots it would be more fair if the three of us chose not to vote. It occurred to me that that might be worth codifying, but maybe not.

TinoDidriksen commented:
> During the most recent election, the election board agreed that since we could see all the ballots it would be more fair if the three of us chose not to vote. It occurred to me that that might be worth codifying, but maybe not.

My gut reaction says no. I'd say the election officials should be able to vote, and with STV their position will be less abusable. And nobody says we have to vote via email - we could use some secure online voting system. Both old and new bylaws allow that.

khannatanmai commented:
> During the most recent election, the election board agreed that since we could see all the ballots it would be more fair if the three of us chose not to vote. It occurred to me that that might be worth codifying, but maybe not.

If the main issue is that after seeing all the votes the election commission has the unique position to vote based on the results and sway the election, a simple solution is that the election board can be the first to cast and codify their votes.

In terms of the election commission being able to access who voted for whom, that's really not much of a problem, but yeah as mentioned, it can be solved by using an online voting system. Or we can develop a basic one ourselves by generating private keys for each voter or something like that.

scoopgracie commented:
Also, when censuring the PMC, it should be possible to censure only one member.

scoopgracie commented:
8fcf39dd03cbc650b9a1453522d5db2216fcd9ed "UK spelling" is this a de facto standard for Apertium? Perhaps we should specify this somewhere (not necessarily the bylaws)?

scoopgracie commented on Bylaws.md:
```diff
@@ -19,92 +19,116 @@
-	<li>Developers are  volunteers who are contributing time, code, documentation, or resources to Apertium. A developer that makes sustained, welcomed contributions to the project may be invited to become a committer, though the exact timing of such invitations depends on many factors</li>
+	<li>Contributors are volunteers who contribute time, code, documentation, or resources to Apertium. A contributor that makes sustained, welcomed contributions to Apertium may be invited to become a Voting Member, though the exact timing of such invitations is not specified.</li>
+	<li>Contributors who have administrative rights to a repository may exercise those rights almost fully, including granting rights to new contributors, making releases, or setting up automation. Potentially destructive or distruptive actions, such as deleting or renaming the repository or master branch, or breaking backwards compatibility, must be approved by a PMC member, who may decide the acti ...
+	<li>Apertium recommends, but does not require, that contributors sign the Apertium Free/Open-Source Contributor Licence Agreement. However, contributions that are not licensed to Apertium in this way must be clearly marked with a GPLv2+ compatible licence, plus the author's name and email or other identifying information sufficient for a lawyer to reach the contributor.</li>
```
There should be a clause either specifying that minors should not sign the CLA, or providing a procedure for a parent or legal guardian to sign it on behalf of the minor.

scoopgracie commented:
Also, I think we should codify that if any Apertium repo/Website/etc. is compromised due to any Contributor's negligence to secure his/her own account, that Contributor should be considered fully responsible. Maybe add an exception for IRC spam, since "hacking" an account and using someone else's nick is so easy, even with nickserv.

scoopgracie commented:
Summarizing my current requests and deleting the comments for cleanup:
* I believe there should be some form of a judicial system to interpret the Bylaws.
* It should be possible to censure one member of the PMC.
* The CLA clause should have a special case for minors.
* If, by a Contributor's negligence to secure his/her accounts, any Apertium repo/Website/etc. is compromised, that Contributor should be held fully responsible. There should be exceptions for IRC and email spam, since nick and address spoofing are trivial for those platforms.

TinoDidriksen commented:
> I believe there should be some form of a judicial system to interpret the Bylaws.

There is. The Assembly. All decisions must be reviewed by the Assembly, who can overturn the PMC. But as I mentioned earlier, I think we should have a long form document that writes out rationale and interpretation for each clause, so that future questions can be prevented.

> It should be possible to censure one member of the PMC.

Again, the Assembly can do that.

> The CLA clause should have a special case for minors.

I don't think that's possible. We can't write in terms that will legally cover minors. Copyright law intersection with minor labour laws across all jurisdictions in the world? I don't even want to try. Better to have a broadly applicable CLA and let each contributor figure out how they can legally sign it, if they want to sign it.

> If, by a Contributor's negligence to secure his/her accounts, any Apertium repo/Website/etc. is compromised, that Contributor should be held fully responsible. There should be exceptions for IRC and email spam, since nick and address spoofing are trivial for those platforms.

That is already covered by the current removal wording. No need to specify further.

khannatanmai commented:
> * If, by a Contributor's negligence to secure his/her accounts, any Apertium repo/Website/etc. is compromised, that Contributor should be held fully responsible. There should be exceptions for IRC and email spam, since nick and address spoofing are trivial for those platforms.

I know these are bylaws or whatever, but there really is no reason to be this harsh. A contributor should be held fully responsible for events that they have the intention to carry out. If someone gets hacked, rather than holding then fully responsible, assistance must be given to secure their accounts. Even as a preventive we can help secure the accounts of our contributors, but to hold them responsible for the actions of a hacker is unnecessarily punitive.

jonorthwash commented:
Just curious: what does the CLA look like, how does signing it work, and how will Apertium record/track/store signings?

TinoDidriksen commented:
> Just curious: what does the CLA look like, how does signing it work, and how will Apertium record/track/store signings?

It's in the PR: https://github.com/apertium/organisation/blob/overhaul/CLA-optional.md

Signing can be as simple as pasting the whole CLA into an email with an extra line akin to "I agree that all my contributions to Apertium are bound by this license", and sending that to apertium-stuff or somewhere else public where we can and collect it. That would be sufficient for Danish law, at least.

To be extra sure, we can add a step where a PMC member replies to the email asking if they really mean it, and also checks that the sender's address is actually one they have used in commits.

scoopgracie commented:
I can confirm that that is binding under American law as well.

On Sat, May 2, 2020, 07:12 Tino Didriksen <notifications@github.com> wrote:

> Just curious: what does the CLA look like, how does signing it work, and
> how will Apertium record/track/store signings?
>
> It's in the PR:
> https://github.com/apertium/organisation/blob/overhaul/CLA-optional.md
>
> Signing can be as simple as pasting the whole CLA into an email with an
> extra line akin to "I agree that all my contributions to Apertium are bound
> by this license", and sending that to apertium-stuff or somewhere else
> public where we can and collect it. That would be sufficient for Danish
> law, at least.
>
> To be extra sure, we can add a step where a PMC member replies to the
> email asking if they really mean it, and also checks that the sender's
> address is actually one they have used in commits.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/apertium/organisation/pull/13#issuecomment-622959767>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AN54XP76HZTGIFZF44HNORTRPQS5NANCNFSM4MH2I5SQ>
> .
>

scoopgracie commented:
Although signing it with a secret key (GitHub SSH key would be fine) would help. Email spoofing is just too easy to trust.

scoopgracie commented:
@TinoDidriksen can we add this?
    When at least half plus one of the members of the Project Management Committee or at least one-tenth of the Assembly deem it necessary, an Amendment to these Bylaws shall be proposed, which, upon the approval of at least two-thirds of the Assembly, shall be valid to all intents and purposes as part of these Bylaws.

scoopgracie commented:
@TinoDidriksen also, can we remove the sponsorship thing for becoming a voting member and just say any contributor can apply for voting member status and the PMC votes on it?

scoopgracie commented:
(To avoid possible plagiarism, the amendment clause I wrote is mostly copied from Article V, Clause 1 of the US Constitution.)

scoopgracie commented:
is it intended that contributors would ask for nomination, or the PMC would do it on their own?
